src/System.Private.CoreLib/src/System/Diagnostics/Eventing/RuntimeEventSource.cs

```diff
@@ -0,0 +1,140 @@
+        {
+            s_RuntimeEventSource = new RuntimeEventSource();
+        }
+
+        // Threads
+        // TODO
+        
+        public RuntimeEventSource(): base(EventSourceSettings.EtwSelfDescribingEventFormat)
+        {
+
+        }
+
+        protected override void OnEventCommand(System.Diagnostics.Tracing.EventCommandEventArgs command)
+        {
+            if (command.Command == EventCommand.Enable)
+            {
+                // TODO: These are all returning fake stuff now
+                _counters = new EventCounter[] {
+                    // process info counters
+                    new EventCounter("Total Process Time", this),
+                    new EventCounter("Working Set", this),
+                    new EventCounter("Handle Count", this),
+                    new EventCounter("Thread Count", this),
+
+                    // GC info counters
+                    new EventCounter("Total Memory by GC", this),
+                    new EventCounter("Gen 0 GC Count", this),
+                    new EventCounter("Gen 1 GC Count", this),
+                    new EventCounter("Gen 2 GC Count", this),
+
+                    // TODO: Expose a managed API for computing this
+                    new EventCounter("Exception Count", this)
+                };
+
+                // Initialize the timer, but don't set it to run.
+                // The timer will be set to run each time PollForTracingCommand is called.
+
+                // TODO: We might not need this timer once we are done settling upon a high-level design for
+                // what EventCounter is capable of doing. Once that decision is made, we might be able to 
+                // get rid of this.
+                _timer = new Timer(
+                    callback: new TimerCallback(PollForCounterUpdate),
+                    state: null,
+                    dueTime: Timeout.Infinite,
+                    period: Timeout.Infinite,
+                    flowExecutionContext: false);
+
+                // Trigger the first poll operation on the start-up path.
+                PollForCounterUpdate(null);
+            }
+            else if (command.Command == EventCommand.Disable)
+            {
+
+                // TODO: There are some multi-threading issues where this can be modified from a OnEventCommand callback
+                // as well as a timer callback, so we need some kind of synchronization. We need to decide on what kind of 
+                // thread safety EventCounter provides to determine what kind of synchronization is needed here.
+
+                // Dispose counters when perfcounters are disabled
+                _counters = null;
+
+                if (_timer != null)
+                {
+                    _timer.Dispose();
+                    _timer = null;
+                }
+            }
+        }
+
+        public void UpdateAllCounters()
+        {
+            // TODO: These are all returning fake stuff for now
```

Comment:
You probably want to cache counter in a local and check it for null to spurious NullReferenceExceptions on race conditions.

Reply:
Agreed there are race conditions here, but I'm proposing we simply note them and ensure the final version of the feature has it resolved. I am anticipating the timer won't exist here at that point.